lib/dialog.js

```diff
@@ -183,18 +189,31 @@ class Dialog extends Emitter {
     }
     this.queuedRequests = [];
 
+    const removeDialog = (err, res, callback) => {
+      this.connected = false ;
+      this.srf.removeDialog(this) ;
+      callback(err, res) ;
+      this.removeAllListeners();
+    };
+
     const __x = (callback) => {
       if (this.dialogType === 'INVITE') {
         this.agent.request({
           method: 'BYE',
           headers: opts.headers || {},
           stackDialogId: this.id,
+          auth: opts.auth || this.auth,
           _socket: this.socket
         }, (err, bye) => {
-          this.connected = false ;
-          this.srf.removeDialog(this) ;
-          callback(err, bye) ;
-          this.removeAllListeners();
+          if (err) {
+            removeDialog(err, bye, callback);
+          }
+
+          bye.on('response', (res) => {
+            if (res.status === 200) {
```

Comment:
We should actually call removeDialog in all cases here, regardless of the status on the response.  If we got a 401/407, that should have been handled at the lower level of drachtio agent, and the response we will get here will be the subsequent 200 OK to our resent BYE with credentials.  

However, we also have to consider the case where we are running this version of drachtio-srf against a drachtio-server that has not been updated to have the associated fix for BYE auth challenge.  In that case, we *will* get a 401/407 at this level, and we should still clear the dialog as there is nothing more we can do.

Reply:
By the way, thanks for adding the new test case for this scenario.  That is awesome!